### PR TITLE
Add handling for custom-routes with basePath

### LINKED ISSUE #14782: Figure out interplay between basePath and custom redirects / rewrites
As the title states, we need to decide the story for basePath & custom routes for the docs examples.

### PR DESCRIPTION
**Pull request by ijjk**
This adds handling for custom-routes with `basePath` to automatically add the `basePath` for custom-routes `source` and `destination` unless `basePath: false` is set for the route. 

Closes: https://github.com/vercel/next.js/issues/14782

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| buildDuration | 9.8s | 9.8s | ⚠️ +8ms |
| nodeModulesSize | 66.4 MB | 66.4 MB | ⚠️ +10.4 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.574 | 1.626 | ⚠️ +0.05 |
| / avg req/sec | 1588.58 | 1537.49 | ⚠️ -51.09 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 0.959 | 0.993 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 2607.18 | 2517.87 | ⚠️ -89.31 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..e4a8.js gzip | 10.7 kB | 10.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.72 kB | 5.72 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 7.1 kB | 7.1 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _buildManifest.js gzip | 268 B | 268 B | ✓ |
| _buildManife..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| index.html gzip | 953 B | 953 B | ✓ |
| link.html gzip | 961 B | 961 B | ✓ |
| withRouter.html gzip | 945 B | 945 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| buildDuration | 10.5s | 10.4s | -62ms |
| nodeModulesSize | 66.4 MB | 66.4 MB | ⚠️ +10.4 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..e4a8.js gzip | 10.7 kB | 10.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.72 kB | 5.72 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 7.1 kB | 7.1 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _buildManifest.js gzip | 268 B | 268 B | ✓ |
| _buildManife..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _error.js | 876 kB | 878 kB | ⚠️ +1.82 kB |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 877 kB | 878 kB | ⚠️ +1.82 kB |
| link.js | 916 kB | 918 kB | ⚠️ +1.82 kB |
| routerDirect.js | 910 kB | 911 kB | ⚠️ +1.82 kB |
| withRouter.js | 910 kB | 911 kB | ⚠️ +1.82 kB |
| Overall change | 4.5 MB | 4.5 MB | ⚠️ +9.09 kB |

</details>

</details>
Commit: 7ffa78286ea95065925e812bff6ae4b844f5c415

**Review by timneutkens — APPROVED** [review]
The documentation needs to be updated to cover this, besides that it looks good 👍

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| buildDuration | 12.4s | 12.7s | ⚠️ +288ms |
| nodeModulesSize | 66.4 MB | 66.4 MB | ⚠️ +10.4 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.232 | 2.228 | 0 |
| / avg req/sec | 1120.2 | 1122.15 | +1.95 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.498 | 1.45 | -0.05 |
| /error-in-render avg req/sec | 1668.7 | 1724.25 | +55.55 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..e4a8.js gzip | 10.7 kB | 10.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.72 kB | 5.72 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 7.1 kB | 7.1 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _buildManifest.js gzip | 268 B | 268 B | ✓ |
| _buildManife..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| index.html gzip | 953 B | 953 B | ✓ |
| link.html gzip | 961 B | 961 B | ✓ |
| withRouter.html gzip | 945 B | 945 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| buildDuration | 13.3s | 13.4s | ⚠️ +37ms |
| nodeModulesSize | 66.4 MB | 66.4 MB | ⚠️ +10.4 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..e4a8.js gzip | 10.7 kB | 10.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57.2 kB | 57.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.72 kB | 5.72 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 7.1 kB | 7.1 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.7 kB | 52.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _buildManifest.js gzip | 268 B | 268 B | ✓ |
| _buildManife..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/custom-routes-base-path-interop | Change |
| - | - | - | - |
| _error.js | 876 kB | 878 kB | ⚠️ +1.82 kB |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 877 kB | 878 kB | ⚠️ +1.82 kB |
| link.js | 916 kB | 918 kB | ⚠️ +1.82 kB |
| routerDirect.js | 910 kB | 911 kB | ⚠️ +1.82 kB |
| withRouter.js | 910 kB | 911 kB | ⚠️ +1.82 kB |
| Overall change | 4.5 MB | 4.5 MB | ⚠️ +9.09 kB |

</details>

</details>
Commit: 3c63b83cad06e58e15c29148cfa4d7e186594c01

File: packages/next/next-server/server/next-server.ts [hunk 1]
```diff
@@ -253,13 +253,11 @@ export default class Server {
...
-      return this.render404(req, res, parsedUrl)
-    } else {
-      // If replace ends up replacing the full url it'll be `undefined`, meaning we have to default it to `/`
-      parsedUrl.pathname = parsedUrl.pathname!.replace(basePath, '') || '/'
-      req.url = req.url!.replace(basePath, '')
+    if (basePath && req.url?.startsWith(basePath)) {
+      // store original URL to allow checking if basePath was
+      // provided or not
+      ;(req as any)._nextHadBasePath = true
+      req.url = req.url!.replace(basePath, '') || '/'
```
Comment (nodkz) [inline packages/next/next-server/server/next-server.ts:260]: @ijjk @timneutkens what about adding `req.baseUrl = basePath;` to [express req object](https://expressjs.com/en/api.html#req.baseUrl) in this code block?

I have a problem that I cannot make a proper `redirect` in `getInitialProps()` with `basePath`. I have no ability to read it from `ctx` or `ctx.req` object on the server-side for proper redirect URL construction.

**Pull request merged**